worker/sshserver/server.go

```diff
@@ -66,48 +77,40 @@ func NewServerWorker(config ServerWorkerConfig) (worker.Worker, error) {
 		},
 	}
 
-	// TODO(ale8k): Update later to use the host key from StateServingInfo
-	terminatingHostKey, err := rsa.GenerateKey(rand.Reader, 2048)
-	if err != nil {
-		return nil, errors.Annotate(err, "failed to generate host key")
-	}
-	signer, err := gossh.NewSignerFromKey(terminatingHostKey)
-	if err != nil {
-		return nil, errors.Annotate(err, "failed to create signer")
+	// Set hostkey.
+	if err := s.setJumpServerHostKey(); err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	// Set hostkey
-	s.Server.AddHostKey(signer)
-
 	if s.config.Listener == nil {
-		listenAddr := net.JoinHostPort("", strconv.Itoa(2229))
+		listenAddr := net.JoinHostPort("", strconv.Itoa(config.Port))
 		listener, err := net.Listen("tcp", listenAddr)
 		if err != nil {
 			return nil, err
 		}
 		s.config.Listener = listener
 	}
 
-	listener := newAcceptOnceListener(s.config.Listener)
+	listener, closeAllowed := NewSSHServerListener(s.config.Listener)
 
-	// Start server
+	// Start server.
 	s.tomb.Go(func() error {
 		err := s.Server.Serve(listener)
 		if errors.Is(err, ssh.ErrServerClosed) {
 			return nil
 		}
-		return err
+		return errors.Trace(err)
 	})
 
-	// Handle server cleanup
+	// Handle server cleanup.
 	s.tomb.Go(func() error {
 		<-s.tomb.Dying()
-		<-listener.closeAllowed
+		<-closeAllowed
```

Comment:
One thing I would do is guard this second with a select+timeout. If the the tomb has been killed, we want to wait a reasonable time for natural attrition, then go ahead and wrap up.

Reply:
Added a timeout of 10 second